src/array.cr

```diff
@@ -2029,6 +2029,36 @@ class Array(T)
     @capacity - @offset_to_buffer
   end
 
+  private def rewind
+    root_buffer.copy_from(@buffer, @size)
+    shift_buffer_by -@offset_to_buffer
+  end
+
+  # Enusures that the internal buffer has at least `capacity` elements. If
+  # `rewind` is truthy, then the internal buffer is rewind first: the
+  # elements are copied to the front of the array.
+  def ensure_capacity(capacity : Int32, rewind = true) : self
+    rewind() if rewind
+
+    if capacity - @offset_to_buffer >= @size
+      resize_to_capacity capacity
+    else
+      err = "Insufficient capacity: #{capacity} cannot hold #{@size} elements"
+      err += " with an offset of #{@offset_to_buffer}" if @offset_to_buffer > 0
+      raise ArgumentError.new(err)
+    end
+    self
+  end
+
+  # Reduces the internal buffer to exactly fit the number of elements in the
+  # array. If `rewind` is truthy it moves the elements in the internal array to the front.
+  def trim_to_size(rewind = true) : self
```

Comment:
I'm not convinced about the usefulness of the `rewind` parameter. `#trim_to_size` should always rewind. I can't see a relevant usecase where you would not want that, especially since you have no external insight on or control of the offset.
For `#ensure_capacity` it's not that clear, but I don't know why it should bother about that either.

Reply:
Agree, without knowing the internals of `Array` no one would know what "rewind" implies

Reply:
would it be useful to have an "extra capacity" argument, that is `resize_to_capacity(size + extra_capacity)`? I think we would want that, so that if an element was added immediately after the trimming, it wouldn't needed to increase capacity so close again. on the other hand i guess `ensure_capacity` can be used for that.

Reply:
Yeah, I think the use case for `#trim_to_size` would be to say I'm done with adding or removing things, let's have it trimmed to save space.

Reply:
As it is implemented here, `ensure_capacity` won't shrink, so it won't help you in that case. This is why my original proposal had just one method to set the capacity, either expanding or shrinking it. I wonder what's best, if to go back to the original proposal or to add an extra-capacity argument to leave some room.

Reply:
I wonder if fixing https://github.com/crystal-lang/crystal/issues/11445 would help here in your use case @carlhoerberg .

Reply:
no, it's a very good addition, but our deques/arrays also shrinks. well, we can continue to do what we do, create a new array/deque and copy all elements to it, but not having to copy would be much more efficent.

Reply:
I would say that we indeed should never expose the `rewind` parameter and always do it.
Later someone can separately research whether that's a significant slowdown, and then possibly add this parameter. Also whether it's perhaps possible to do a 2-in-1 equivalent of `realloc` + `rewind`.

Reply:
But, then again, potentially deprecating the `rewind` parameter to be ignored (always true) would also not be a breaking change. So maybe it's fine as is.

Reply:
Please let's not add the rewind parameter. Nothing in Array's API says anything about the internal implementation. This will prevent changing the internal implementation later on. Also users reading this will have no idea what this means.